pkg/provisioner/ironic/ironic.go

```diff
@@ -1368,3 +1371,107 @@ func (p *ironicProvisioner) softPowerOff() (result provisioner.Result, err error
 
 	return result, nil
 }
+
+// Describes the current availability status of a generic dependency
+// required by the provisioner to operate correctly
+type provisionerDependencyStatus struct {
+	// The name of the dependency
+	name string
+	// If ready, the dependency is available to process incoming requests
+	ready bool
+	// Can provide additional info if the dependency is not ready
+	err error
+}
+
+// Check to verify dependency status
+type provisionerDependencyCheck func(chan<- provisionerDependencyStatus)
+
+// IsReady checks if the provisioning backend is available
+func (p *ironicProvisioner) IsReady() (bool, error) {
+	p.log.Info("verifying ironic provisioner dependencies")
+
+	checks := []provisionerDependencyCheck{
+		p.checkIronic,
+		p.checkInspector,
+	}
+
+	var done = make(chan provisionerDependencyStatus, len(checks))
+	defer close(done)
+
+	for _, check := range checks {
+		go check(done)
+	}
+
+	results := []provisionerDependencyStatus{}
+	for i := 0; i < len(checks); i++ {
+		results = append(results, <-done)
+	}
+
+	for _, r := range results {
+		if !r.ready {
+			p.log.Info("dependency not yet ready", "name", r.name)
+			return r.ready, r.err
+		}
+	}
+
+	return true, nil
+}
+
+func (p *ironicProvisioner) checkIronic(done chan<- provisionerDependencyStatus) {
+	status := p.checkEndpoint(p.client, "ironic")
+	if status.ready {
+		status = p.checkConductor()
+	}
+	done <- status
+}
+
+func (p *ironicProvisioner) checkInspector(done chan<- provisionerDependencyStatus) {
+	done <- p.checkEndpoint(p.inspector, "ironic-inspector")
+}
+
+func (p *ironicProvisioner) checkEndpoint(client *gophercloud.ServiceClient, serviceName string) (status provisionerDependencyStatus) {
+
+	status.name = serviceName
+
+	httpClient := &http.Client{
+		Timeout: 10 * time.Second,
+	}
+
+	// NOTE: Some versions of Ironic inspector returns 404 for /v1/ but 200 for /v1,
+	// which seems to be the default behavior for Flask. Remove the trailing slash
+	// from the client endpoint.
+	endpoint := strings.TrimSuffix(client.Endpoint, "/")
+
+	resp, err := httpClient.Get(endpoint)
+	status.err = err
+	if status.err == nil {
+		status.ready = resp.StatusCode == http.StatusOK
+		resp.Body.Close()
```

Comment:
Even when there is no error, would we want to call resp.Body.Close()?

Reply:
Good tip, let me recheck it

Reply:
The pattern is fine, only in case of no error it must be invoked to avoid resource leaks